### PR TITLE
Implement contract instance storage in Soroban host.

### PR DESCRIPTION
**Pull request by dmkozh**
### What

Implement contract instance storage in Soroban host.

Also use instance storage in built-in token contract for instance data (metadata, admin, asset info).

### Why

Instance storage allows storing contract configuration for cheaper and also simplifies its lifetime management - https://github.com/stellar/rs-soroban-env/issues/860

### Known limitations

N/A

### CODE REVIEW

File: soroban-env-host/src/vm.rs [hunk 1]
```diff
@@ -281,85 +276,38 @@ impl Vm {
...
-    pub fn functions(&self) -> Vec<VmFunction> {
-        let mut res = Vec::new();
-        for e in self.module.exports() {
-            if let wasmi::ExternType::Func(f) = e.ty() {
-                res.push(VmFunction {
-                    name: e.name().to_string(),
-                    param_count: f.params().len(),
-                    result_count: f.results().len(),
-                })
+        let wasm_args: Vec<Value> = args
```
Comment (dmkozh) [inline soroban-env-host/src/vm.rs:279]: @jayz22 Just FYI, `InvokeVmFunction` has to be recalibrated with this change - previously it used to measure charging budget for itself, as well as creating a guard frame (which is metered/calibrated separately).

File: soroban-env-host/src/host.rs [hunk 2]
```diff
@@ -264,6 +269,53 @@ impl Host {
...
     }
 
+    /// Immutable accessor to the instance storage of the currently running
+    /// contract.
+    /// Performs lazy initialization of instance storage on access.
+    pub fn with_instance_storage<F, U>(&self, f: F) -> Result<U, HostError>
+    where
+        F: FnOnce(&InstanceStorageMap) -> Result<U, HostError>,
+    {
+        // f(&mut self.0.storage.borrow_mut())
```
Comment (sisuresh) [inline soroban-env-host/src/host.rs:279]:
```suggestion
```

File: soroban-env-host/src/vm.rs (hunk 1 above)
Comment (jayz22) [reply soroban-env-host/src/vm.rs:279]: Cool, I'll worry about it later.

File: soroban-env-host/src/cost_runner/cost_types/guard_frame.rs [hunk 3]
```diff
@@ -12,10 +14,15 @@ impl CostRunner for GuardFrameRun {
     type RecycledType = Option<Self::SampleType>;
 
     fn run_iter(host: &crate::Host, _iter: u64, sample: Self::SampleType) -> Self::RecycledType {
+        let empty_instance = ScContractInstance {
+            executable: ContractExecutable::Token,
+            storage: None,
+        };
```
Comment (jayz22) [inline soroban-env-host/src/cost_runner/cost_types/guard_frame.rs:20]: You don't have to anything now, just wanted to mention it. 
The right way to do it is to pass the `empty_instance` from the `sample`, and declare the additional field in the `SampleType`.
The sample set up is done in `benches/common/cost_types/guard_frame.rs` (same name different directory). 
This way we minimize the noise of the setup.

File: soroban-env-host/src/storage.rs [hunk 4]
```diff
@@ -19,6 +19,20 @@ use crate::{host::metered_map::MeteredOrdMap, HostError};
 
 pub type FootprintMap = MeteredOrdMap<Rc<LedgerKey>, AccessType, Budget>;
 pub type StorageMap = MeteredOrdMap<Rc<LedgerKey>, Option<Rc<LedgerEntry>>, Budget>;
+#[derive(Clone)]
+pub struct InstanceStorageMap {
+    pub map: MeteredOrdMap<Val, Val, Host>,
```
Comment (jayz22) [inline soroban-env-host/src/storage.rs:24]: Can this be changed to a `Budget` instead of `Host`? As far as I can tell this doesn't need anything else from the Host besides the budget. This way it can be used outside of the host (e.g. setting up a storage without the host).

File: soroban-env-host/src/vm.rs [hunk 5]
```diff
@@ -281,85 +276,38 @@ impl Vm {
...
-        let mut raw_args: Vec<Val> = Vec::new();
-        for scv in args.0.iter() {
-            raw_args.push(host.to_host_val(scv)?);
-        }
-        let raw_res = self.invoke_function_raw(host, &func_sym, raw_args.as_slice())?;
-        host.from_host_val(raw_res)
-    }
-
-    /// Returns a list of functions in the WASM module loaded into the [Vm].
-    pub fn functions(&self) -> Vec<VmFunction> {
```
Comment (jayz22) [inline soroban-env-host/src/vm.rs:351]: Why did you remove this function? 
I don't know if it's being used but theoretically it could be used by e.g. the sdk/CLI for displaying some useful info.

File: soroban-env-host/src/storage.rs (hunk 4 above)
Comment (dmkozh) [reply soroban-env-host/src/storage.rs]: I don't think so - this stores `Val` and `Val` is meaningless outside of the host (that's why we don't have comparison implemented for `<Val, Budget>`. Moreover, this is deeply internal data structure and it shouldn't be used oustside of the host (I'll actually switch it to pub(crate)).

File: soroban-env-host/src/cost_runner/cost_types/guard_frame.rs (hunk 3 above)
Comment (dmkozh) [reply soroban-env-host/src/cost_runner/cost_types/guard_frame.rs]: Updated.

File: soroban-env-host/src/vm.rs (hunk 5 above)
Comment (dmkozh) [reply soroban-env-host/src/vm.rs:351]: Nobody seems to be using it. SDK and CLI use the spec to generate typed function specs (this will just return functions that operate on `Val`s, which isn't super useful). Also, I think we shouldn't expose any vm functions - if we ever need to get info for it, we'd rather do that via host interface (VM is too tightly coupled to the host now, so using it outside of the host leads to footguns, like the `invoke_function` issue I've encountered after refactoring.

File: soroban-env-host/src/cost_runner/cost_types/guard_frame.rs (hunk 3 above)
Comment (jayz22) [reply soroban-env-host/src/cost_runner/cost_types/guard_frame.rs]: Great, thanks!

File: soroban-env-host/src/storage.rs (hunk 4 above)
Comment (jayz22) [reply soroban-env-host/src/storage.rs]: Ok, makes sense.

File: soroban-env-host/src/vm.rs (hunk 5 above)
Comment (jayz22) [reply soroban-env-host/src/vm.rs:351]: Sounds good. I'm not objecting to removing. Was just suggesting to double check.

**Review by graydon — APPROVED** [review]
